### PR TITLE
Addon files

### LINKED ISSUE #384: generate checksums for all Components files [$15]
It would be amazing if aptly would generate [AppStream](https://www.freedesktop.org/wiki/Distributions/AppStream/) data when it publishes a repository. This would allow packages in the published repo to be seen by new-age app centers, including gnome software.

[Here](https://github.com/ximion/appstream-generator) is a generator written in D for reference.

### PR DESCRIPTION
**Pull request by btkostner**
Fixes #384 

## Description of the Change

This is a follow up  / redo of my #420 appstream PR. I realized I made the previous PR too narrow, and it touched a lot of code that it did not need to. Here is what this does:

- Adds an `addon` folder at `.aptly/addon` that can store files you want to be in the repository
- On publish, aptly will look in that folder and write the files to the repository, and release files
- I did not add any API or CLI tags to change the addon directory, as that would be bad security
- All files will be copied AS IS. No compressing or signing

The addon folder mirrors the public folder, so on publish `.aptly/addon/dists/maverick/main/dep11/Components-i386.yml.gz` will create `.aptly/public/dists/maverick/main/dep11/Components-i386.yml.gz` or equivalent.

## Checklist

- ~unit-test added (if change is algorithm)~
- [x] functional test added/updated (if change is functional)
- ~man page updated (if applicable)~
- ~documentation updated~
- [X] author name in `AUTHORS`

### CODE REVIEW

**Comment by btkostner** [comment]
I have a couple of questions before this is all complete about the checklist:

- All of the functions will be tested with functional tests. Do you still want unit tests for the trivial golang functions?
- I wrote a basic test in python, but does not seem to work. I'm guessing because the file is getting removed before the test. Any chance I could get some help on this one?
- Since I didn't add any CLI or API tags, is there anything you want me to add to the man pages?
- Is the documentation all on [the website](https://github.com/smira/aptly-dev.github.io)?

Thanks you!

**Comment by smira** [comment]
> All of the functions will be tested with functional tests. Do you still want unit tests for the trivial golang functions? Most of it's file system work, which would be hard to test without some weird mocking.

I don't think we need unit-tests for things covered with functional testing. I prefer functional tests, but some things are hard to test (e.g. retry algorithm). So if covered with functional, we don't need any unit-tests.

**Comment by btkostner** [comment]
I added a functional test that:
- creates a simple test files
- ensure it's published when aptly publishes a repository
- ensure the checksum is in release file

Let me know if there is anything else for this you want covered.

**Comment by janisppc** [comment]
As this is still open are there any future plans for this AppStream (addon files) feature?

**Comment by smira** [comment]
@btkostner @janisppc I'll take a look tomorrow

**Comment by btkostner** [comment]
This should be up to date, and ready to merge now

File: deb/publish.go [hunk 1]
```diff
@@ -485,9 +485,49 @@ func (p *PublishedRepo) GetLabel() string {
...
+// path from component folder, and value being the full local FS path.
+func (p *PublishedRepo) GetAddonFiles(addonDir string, component string) (map[string]string, error) {
+	files := make(map[string]string)
+
+	if addonDir == "" {
+		return files, nil
+	}
+
+	fsPath := filepath.Join(addonDir, p.Prefix, "dists", p.Distribution, component)
+	if err := filepath.Walk(fsPath, func(path string, info os.FileInfo, err error) error {
```
Comment (apachelogger) [inline deb/publish.go]: If we use Walk here I think it also needs to implement some level of symlink handling. I can entirely imagine people wanting to keep the data in some other dir and then `ln -sf` resources it into the addon dir as they get updated.

**Comment by apachelogger** [comment]
@btkostner If you could rebase with master that'd be lovely, and maybe have a look at the symlink walking and documentation.

We are also having some concerns over the name. Addon makes it sound like it's some sort of addon system for aptly in general. @sliverc suggested `skel` as a possible name, which I find fairly reasonable. Alternatively I'd offer up `PublicAddonData`.

I'd love to get this merged soon :tm:

**Comment by btkostner** [comment]
Test failure is unrelated to this PR

**Comment by btkostner** [comment]
@apachelogger Updated and renamed to `skel` in the code and `skeleton files` in some of the docs.

As far as the symlink walking, that is going to be a lot of extra work. As far as I know (but it's been a long time since I've touched go so I could be wrong) there is no standard way of handling symlinks.

**Comment by btkostner** [comment]
Opened up a basic doc PR. It might need more information.

**Comment by apachelogger** [comment]
The system tests will make me go insane some day -.-

@btkostner Thanks. You are absolutely right about the symlink, I misremembered how `Walk` works. We'd have to manually readdir and descend into symlinks, so let's just ignore this until someone actually complains.

I'll see what can be done about the failing system test.

**Comment by rhertzog** [comment]
@btkostner What't the status of this feature on your side? It would be really nice if we could have this working in aptly. Missing appstream metadata is a problem right now for us. I might be able to help with some sponsorship to get this completed.

**Comment by btkostner** [comment]
@rhertzog Sadly I have no time to work on this. If I remember correctly, the only thing that needs to be fixed / added is uncompressed checksums not being generated.

**Comment by silopolis** [comment]
This would be a really nice addition and probably make Aptly stand out of the crowd, if not already ;)
@rhertzog I'd modestly but gladly tip this :)

**Comment by silopolis** [comment]
Le dim. 10 juil. 2022 à 00:44, Benj Fassbind ***@***.***> a
écrit :

> Reopened #473 <https://github.com/aptly-dev/aptly/pull/473>.
>

Thanks for re-opening this one!

The world of modern desktop software management apps users will hopefully
soon be grateful... 🙏🙂

**Comment by jhonny-oliveira** [comment]
Is there any chance this ever gets finalized and merged?
I don't mind testing it.

Thank you!

**Comment by silopolis** [comment]
@btkostner @rhertzog any estimate of the work needed and potential funding needed to get this merged?

**Comment by btkostner** [comment]
I'm not going to lie, it's been half a decade since I've looked at this code. Considering how long this has been opened, it's probably going to be easier to use some of this code as reference and open a whole new PR

**Comment by randombenj** [comment]
@reglim could you have a look at this? Lets discuss this next week :)

**Comment by jhonny-oliveira** [comment]
I would like to add that Gnome Software and Ubuntu Software Center require dep11 metadata to display the software, otherwise it won't display it. In addition, I did not manage to find other repository management tool that automatically generates/incorporates this data. So, having this functionality working on aptly would be really welcome.

If I got this correctly (correct me, if I'm wrong), the intention of this pull request was to merge pre-created dep11 data from the addon folder during the publish stage. Conceptually the idea seems quite straight forward and has the advantage of offloading the responsibility of creating the dep11 metadata away from aptly. Hopefully this also makes it easier to implement.

As I mentioned earlier, I don't mind testing it, documenting and making the new version available through [XtraDeb.net](https://xtradeb.net) for early adopters.

Cheers!

**Comment by reglim** [comment]
I just migrated your changes from `btkostner:addon-files` to the branch [feature/384-generate-checksums-for-component-files](https://github.com/aptly-dev/aptly/tree/feature/384-generate-checksums-for-component-files). It's now up to date with the current master.

**Comment by randombenj** [comment]
Let's close this in favour of: https://github.com/aptly-dev/aptly/pull/1186

**Pull request closed without merging**